Qualification tool: Add more information for unsupported operators

nartal1 opened the pull request:
This PR fixes #626 and fixes #656 

In 626 - Output file had duplicate values once for Exec and another for Expressions. 
```
Exec" "Filter" ""
"Exec" "Filter" "Filter Exec is not supported as expressions are not supported -  `AtLeastNNulls`"
```
This PR deduplicates the Exec in output files. i.e if an Exec is not supported due to unsupported expressions, then output will contain Exec name along with unsupported expressions. This PR also identifies UDF and dataset operations and includes that as the reason for the unsupported Exec. Also adds a new column "Ignore Operator" which indicates whether the operator is something like a DDL, Table operation which can be ignored. This helps to make sure we document all the unsupported operators in the output file but we can filter the operators to be prioritized based on "Ignore Operator" column.

Sample Output:
```
App ID,Unsupported Type,Details,Notes,Ignore Operator
"application_1700986078276_0001","Exec","LocalTableScan","","true"
"application_1700986078276_0001","Exec","AdaptiveSparkPlan","","true"
"application_1700986078276_0001","Exec","Scan hive cust2cc_v1.stats","","false"
"application_1700986078276_0001","Exec","SetCatalogAndNamespace","","true"
"application_1700986078276_0001","Exec","Execute CreateDataSourceTableAsSelectCommand","","true"
"application_1700986078276_0001","Exec","Filter","Filter Exec is not supported as this operation is from dataset which is not supported","false"
"application_1700992056158_0001","Exec","BatchEvalPython","BatchEvalPython Exec is not supported as it contains UDF which is not supported","false"
"application_1700992056158_0001","Exec","Filter","Filter Exec is not supported as it contains UDF which is not supported","false"
"application_1700992056158_0001","Exec","Project","Project Exec is not supported as expressions are not supported -  `normalizeIp`","false"
"application_1700992056158_0001","Exec","Filter","Filter Exec is not supported as expressions are not supported -  `validateIP;normalizeIp`","false"
"application_1700992056158_0001","Exec","SortMergeJoin","SortMergeJoin Exec is not supported as expressions are not supported -  `normalizeEmail`","false"
"application_1700992056158_0001","Exec","Sort","Sort Exec is not supported as expressions are not supported -  `normalizeEmail`","false"
"application_1700992056158_0001","Expression","normalizeName","","false"
"application_1700992056158_0001","Expression","normalizeEmail","","false"

```

For 656 - Generates new output file (rapids_4_spark_qualification_output_unsupportedOperatorsStageDuration.csv)  based on comment - https://github.com/NVIDIA/spark-rapids-tools/issues/656#issuecomment-1810339041 .  This file helps to determine wallclockStageDuration to appDuration ratio and corresponding unsupported operators in that stage. This helps in prioritizing the unsupported opeators.

Sample output:
```
App ID,Unsupported Type,Stage ID,Stage Duration,App Duration,Recommendation,Ignore Operator
"application_1700992056158_0001","LocalTableScan",65,4323,1691909,Recommended,true
"application_1700992056158_0001","Execute InsertIntoHiveTable",107,3940,1691909,Recommended,false
"application_1700992056158_0001","BatchEvalPython",81,141962,1691909,Recommended,false
"application_1700992056158_0001","Filter",81,141962,1691909,Recommended,false
"application_1700992056158_0001","BatchEvalPython",81,141962,1691909,Recommended,false
"application_1700992056158_0001","Project",81,141962,1691909,Recommended,false
"application_1700992056158_0001","Filter",7,77,1691909,Recommended,false
"application_1700992056158_0001","LocalTableScan",7,77,1691909,Recommended,true
```





<!--

Thank you for contributing to Tools of RAPIDS Accelerator for Apache Spark!

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

4. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

5. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present).

6. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please avoid
   rebasing your branch during the review process, as this causes the context
   of any comments made by reviewers to be lost. If conflicts occur during
   review then they should be resolved by merging into the branch used for
   making the pull request.

Many thanks in advance for your cooperation!

-->

parthosa commented on core/src/main/scala/com/nvidia/spark/rapids/tool/qualification/QualOutputWriter.scala:
```diff
@@ -161,6 +161,23 @@ class QualOutputWriter(outputDir: String, reportReadSchema: Boolean,
+      val rows = QualOutputWriter.constructUnsupportedStagesDurationInfo(sum, headersAndSizes,
+        QualOutputWriter.CSV_DELIMITER, false)
+      rows.foreach(row => csvFileWriter.write(row))
+    }
```
Should we call `csvFileWriter.close()` at the end of this function?

Similar change in `writeUnsupportedOperatorsCSVReport()`.

nartal1 commented on core/src/main/scala/com/nvidia/spark/rapids/tool/qualification/QualOutputWriter.scala:
```diff
@@ -895,81 +925,106 @@ object QualOutputWriter {
-        }
-        val data = ListBuffer(
+    sumInfo.stageInfo.collect {
+      case info if info.stageWallclockDuration > 0 && info.unsupportedExecs.nonEmpty =>
```
It is a redundant check which should have been removed. If we have some unsupported operators then there will be some wallclock duration associated with that stage(unless the stage itself is not completed which is handled separately). PTAL

parthosa reviewed (dismissed):
Thanks @nartal1.

nartal1 commented:
Thanks @parthosa for the review. I updated the PR based on @viadea 's offline feedback. I have updated the description. PTAL.
@viadea - Would be good to get your feedback.

amahussein reviewed (approved):
Thanks @nartal1
